pymc_experimental/statespace/filters/distributions.py

```diff
@@ -0,0 +1,121 @@
+import numpy as np
+import pytensor
+from pymc.distributions.distribution import Distribution, SymbolicRandomVariable
+from pymc.distributions.shape_utils import get_support_shape_1d
+from pymc.logprob.abstract import _logprob
+from pytensor.graph.basic import Node
+from pytensor.tensor.random.op import RandomVariable
+from scipy import stats
+
+
+class _KalmanFilterBaseRV(RandomVariable):
+    name = "_kalmanfilterbase"
+    ndim_supp = 0
+    ndims_params = [2, 2, 2, 2, 2, 2, 2, 2, 2]
+    dtype = "floatX"
+    _print_name = ("_kalmanfilterbase", "\\operatorname{_kalmanfilterbase}")
+
+    def make_node(self, rng, x0, P0, c, d, T, Z, R, H, Q, steps, size):
+        return super().make_node(rng, x0, P0, c, d, T, Z, R, H, Q, steps, size)
+
+    def rng_fn(self, rng, x0, P0, c, d, T, Z, R, H, Q, steps, size=None):
+        n_obs, n_states = Z.shape[:2]
+        n_posdef = Q.shape[0]
+
+        state_innovations = stats.multivariate_normal(mean=np.zeros(n_posdef), cov=Q, seed=rng).rvs(
+            steps
+        )
+        measurement_error = stats.multivariate_normal(
+            mean=np.zeros(n_obs), cov=H, seed=rng, allow_singular=True
+        ).rvs(steps)
+
+        state_innovations = state_innovations[:, :, None]
+        measurement_error = measurement_error[:, None, None]
+
+        hidden_states = np.zeros((steps, n_states, 1))
+        observed = np.zeros((steps, n_obs, 1))
+        hidden_states[0] = x0
+        observed[0] = Z @ x0 + measurement_error[0]
+
+        for t in range(1, steps):
+            hidden_states[t] = T @ hidden_states[t - 1] + R @ state_innovations[t] + c
+            observed[t] = Z @ hidden_states[t] + measurement_error[t] + d
+
+        return observed
+
+
+#         if size is not None:
+#             raise NotImplementedError()
+
+#         (filtered_states, predicted_states, filtered_cov,
+#             predicted_cov, log_likelihood) = kalman_filter(data, x0, P0, c, d, T, Z, R, H, Q)
+
+#         return filtered_states, predicted_states, filtered_cov, predicted_cov
+
+
+_kalman_filter_base = _KalmanFilterBaseRV()
```

Comment:
If you implemented a `RandomVariable` you can skip the `SymbolicRandomVariable` / `SymbolicDistribution` altogether. The latter are just for when you want the random method defined directly with PyTensor expressions.

In `Disttribution` just set that to `rv_op` (instead of `rv_type`). Check how `pm.Normal` is defined

Reply:
I'm trying to work towards a random variable; in that spirit I'm re-writing these filtering functions to explicitly return random variables instead of tensors of parameters. I think I am making some mistakes with the updates though:

```python
from pymc_statespace.filters.kalman_filter import MVN_CONST, BaseFilter
from pymc.pytensorf import collect_default_updates

def get_loc_scale_from_distribution(d):
    *_, loc, scale = d.owner.inputs
    return loc, scale

def make_Z_matrix(y):
    is_nan = pt.isnan(y)
    missing_idx = pt.nonzero(is_nan, return_matrix=False)[0]
    obs_idx = pt.nonzero(pt.neq(is_nan, np.array(True)), return_matrix=False)[0]
    
    n_missing = missing_idx.shape[0]
    n_obs = obs_idx.shape[0]
    n = n_missing + n_obs
    
    Z = pt.set_subtensor(pt.zeros((n_obs, n))[pt.arange(n_obs), obs_idx], 1)
    
    return Z, obs_idx

class PyMCFilter(BaseFilter):
    @staticmethod
    def predict(dist, c, T, R, Q) -> MvNormal:
        a, P = get_loc_scale_from_distribution(dist)
        a_hat = T.dot(a) + c
        P_hat = matrix_dot(T, P, T.T) + matrix_dot(R, Q, R.T)

        # Force P_hat to be symmetric
        P_hat = 0.5 * (P_hat + P_hat.T)
        d_predicted = pm.MvNormal.dist(mu=a_hat, cov=P_hat)
        
        return d_predicted, collect_default_updates(inputs=[dist, c, T, R, Q], outputs=[d_predicted])

    def update(self, y, a, P, d, H):
        """
        Conjugate update rule for the mean and covariance matrix, with log-likelihood en passant
        TODO: Verify these equations are correct if there are multiple endogenous variables.
        TODO: Is there a more elegant way to handle nans?
        """
        Z, obs_idx = make_Z_matrix(y)
        v = y[obs_idx] - Z.dot(a) - d

        PZT = P.dot(Z.T)
        F = Z.dot(PZT) + H
        F_inv = pt.linalg.solve(F, self.eye_endog, assume_a="pos", check_finite=False)
        
        K = PZT.dot(F_inv)
        I_KZ = self.eye_states - K.dot(Z)

        a_filtered = a + K.dot(v)
        P_filtered = matrix_dot(I_KZ, P, I_KZ.T) + matrix_dot(K, H, K.T)

        inner_term = matrix_dot(v.T, F_inv, v)
        ll = -0.5 * (MVN_CONST + pt.log(pt.linalg.det(F)) + inner_term).ravel()[0]
        d_filtered = pm.MvNormal.dist(a_filtered, P_filtered)
        
        return (d_filtered, ll), collect_default_updates(inputs=[y, a, P, d, H], outputs=[d_filtered, ll])
    
    def kalman_step(self, *args):
        """
        The timing convention follows [1]. a0 and P0 are taken to be predicted states, so we begin
        with an update step rather than a predict step.
        References
        ----------
        .. [1] Durbin, J., and S. J. Koopman. Time Series Analysis by State Space Methods.
               2nd ed, Oxford University Press, 2012.
        """
#         y, predicted_dist, c, d, T, R, H, Q = self.unpack_args(args)
        y, a, P, c, d, T, R, H, Q = args
        (filtered_dist, ll), filter_updates = self.update(y=y, a=a, P=P, d=d, H=H)
        
        new_predicted_dist, pred_updates = self.predict(dist=filtered_dist, c=c, T=T, R=R, Q=Q)
        new_a, new_P = get_loc_scale_from_distribution(new_predicted_dist)
        outputs = new_a, new_P, filtered_dist, new_predicted_dist, ll
        
        updates = collect_default_updates(inputs=args, outputs=outputs)
        updates.update(filter_updates)
        updates.update(pred_updates)
    
        return outputs, updates
```

Here's a simple test with `pm.CustomDist` that throws errors about the rng having multiple clients, that's why I think there's a problem with the updates:

```python
p, m, r, n = 1, 5, 2, 10

kf = PyMCFilter()
kf.eye_endog = pt.eye(p)
kf.eye_states = pt.eye(m)

with pm.Model() as mod:
    c = pm.Normal('c', size=(m))
    d = pm.Normal('d', size=(p))
    T = pm.Normal('T', size=(m, m))
    R = pm.Normal('R', size=(m, r))
    H = pm.Normal('H', size=(p, p))
    H = H @ H.T
    Q = pm.Normal('Q', size=(r, r))
    Q = Q @ Q.T
    
    a0 = pm.Normal('a0', size=(m, ))
    P0_sigma = pm.Exponential('P0_sigma', 1)
    P0 = pt.eye(m) * P0_sigma
    
    def kf_func_factory(data, a0, P0, c, d, T, R, H, Q, size=None):
        (*_, filtered_dist, predicted_dist, ll), updates = pytensor.scan(kf.kalman_step,
                             sequences=[data],
                             non_sequences=[c, d, T, R, H, Q],
                             outputs_info=[a0, P0, None, None, None])
        
        def kf_filtered(data, c, d, T, R, H, Q, size=None):
            return filtered_dist
        
        def kf_predicted(data, c, d, T, R, H, Q, size=None):
            return predicted_dist
        
        def kf_logp(data, c, d, T, R, H, Q, size=None):
            return ll
            
        return kf_filtered, kf_predicted, kf_logp
    
    kf_filtered, kf_predicted, kf_logp = kf_func_factory(data, a0, P0, c, d, T, R, H, Q)
    
    data = pytensor.shared(np.c_[np.arange(n)[:, None], np.full((n, m-p), np.nan)], name='data')
    filtered_dist = pm.CustomDist('kalman_filtered', c, d, T, R, H, Q, dist=kf_filtered, logp=kf_logp, observed=data)
    predicted_dist = pm.CustomDist('kalman_predicted', data, c, d, T, R, H, Q, dist=kf_predicted)
```

This raises errors like `C:\Users\Jesse\miniconda3\envs\econ\lib\site-packages\pymc\pytensorf.py:1056: UserWarning: RNG Variable RandomGeneratorSharedVariable(<Generator(PCG64) at 0x193D8EFF220>) has multiple clients. This is likely an inconsistent random graph.`. It also doesn't sample, but one step at a time. Obviously this would all go into an actual `RandomVariable`, I'm just trying to wrap my head around what needs to be done and how it would look.

Reply:
The error says you have a RNG variable that is being used in (at least) two distinct RandomVariables. If you can isolate a MWE I can try to pinpoint the problem more exactly.

Anyway if you put everything inside a single RandomVariable Op you don't have to worry about this sort of stuff. You should have a well defined Op that takes one RNG and does everything in numpy code and returns a well defined set of outputs.

Reply:
You think I should do all this pytensor machinery in numpy? I'm trying not to duplicate code. The pytensor scan could be used to compute the log likelihood, but computing the log lilkelihood always generates the sampling distributions at the same time, so I was trying to think of a way to run the scan, then send the different components (2 distributions + logp graph) to where they need to go.

Basically the problem I'm having is that everything is being done all at once in this one scan, so I am struggling with how to "untangle" it. Another problem is how to untangle the data, which should be passed as observed, from the scanning, which always needs to be done over the data.